docs/intro/settings.rst
**************
Using settings
**************

.. seealso:: :doc:`Full list of settings </reference/settings>`

The ``'settings'`` module in the :file:`/scripts/settings.js` can be used to customize and configure various Nebula
modules in a single place. other modules will reference this settings module and treat it as a central point of a run
time configuration.

::

   define('settings', {
       // Used by require.js, defaults to true, used to toggle 'no conflict mode' for jQuery.
       'JQUERY_NO_CONFLICT': false,

       // Other custom, user-defined settings example:
       'DEBUG': true,
       'ANIMATIONS': true,
       'FPS': 60,
       ...
   });

All settings are considered constants, thus their names are written using upper-case notation.

You can also use a function to generate settings::

   define('settings', function() {
       // Some block of code that calculates various settings.
       var value = someFunction(...);
       ...

       return {
           'SOME_SETTING': value,
           ...
       };
   });


Retrieving settings
===================

Although ``'settings'`` module is usually a simple JavaScript object, when retrieving configuration from other modules,
Nebula gives you a getter function to ease development.

::

   var settings = require('settings');

   if (settings('DEBUG', false))
       ...

Or::

   define('my_module', ['settings'], function(settings) {

       if (settings('DEBUG', false))
           ...

   });

``settings`` in this case is a function. It returns value of a configuration option and in case of an unknown, not
defined setting name, this function returns a fallback value (the second argument).

.. function:: settings(name, fallback)

   Returns value of a setting from the ``'settings'`` module or ``fallback`` if setting with this name is not defined.

   :param string name: Name of a setting.
   :param fallback: Value to return if given setting was not found.


docs/reference/settings.rst
********
Settings
********


Core settings
=============

.. data:: JQUERY_NO_CONFLICT

   Default: ``true``

   Toggles strict `jQuery 'no conflict' mode <http://api.jquery.com/jQuery.noConflict/>`_. By default, when this setting
   is set to ``true``, there are no global ``$`` and ``jQuery`` variables. To use jQuery you need to import it like
   any other JavaScript module, which looks like this::

      var jQuery = require('jquery');

   Or this::

      define('my_module', ['jquery'], function(jQuery) {
          ...
      });

   Note that module names are all lowercase!


Viewport
========

.. data:: VIEWPORTS

   Default: ``{}`` (empty object, used when setting is not set)

   Default supplied in the ``settings.js`` file::

      'VIEWPORTS': {
          '320': [0, 479],
          '480': [480, 719],
          '720': [720, 959],
          '960': [960, 1279],
          '1280': [1280, 1599],
          '1600': [1600, 1919],
          '1920': [1920, Infinity]
      }

   An object whose contents maps viewport aliases to a viewport specification, which is usually an array holding
   minimum and maximum window dimensions.

   ::

      'VIEWPORTS': {
          '(alias)': (specification),
          ...
      }

   ``(alias)`` can be any string value.

   ``(specification)`` can be one of these:

   *  Two-element array: ``[(minWidth), (maxWidth)]`` - minimum and maximum window width.

   *  A callback function that returns ``true`` when given viewport should be activated or ``false`` if not. Such
      function may be used to implement more complex scenarios.

   *  Media query string, e.g. ``'(min-width: 640px) and (max-width: 1280px)'``. Note that media queries are not
      supported in some browsers, including IE8.


.. data:: VIEWPORT_CLASS_PREFIX

   Default: ``'viewport-'``

   A string that will be prepended to the body's class denoting current viewport orientation. By default Nebula uses
   ``viewport-landscape`` and ``viewport-portrait`` classes for this purposes.


.. data:: VIEWPORT_ORIENTATION_CLASS

   Default: ``true``

   Whether to apply ``viewport-lanscape`` or ``viewport-portrait`` class to the body element denoting browser's window
   dimensions.


Smart blocks
============

.. data:: SMART_BLOCKS

   Default: ``{}`` (empty object)

   An object whose contents maps CSS selectors to class names and rules required to apply that class.

   Example::

      'SMART_BLOCKS': {
          'div.block': {
              'small': [0, 100],
              'medium': [101, 200],
              'big': [201, Infinity]
          },
          'ul.grid>li': {
              'narrow': [200, 480, 'self'],
              'wide': [481, Infinity, 'self'],
          }
      }

   General pattern is as follows::

      'SMART_BLOCKS': {
          '(CSS-selector)': {
              '(class-name)': [(min-width), (max-width), ('self')]
              ...
          }
      }

   ``(CSS-selector)``
      This can be any selector supported by a browser.

   ``(class-name)``
      This must be a string with name of the class that will be applied to the element specified with ``(CSS-selector)``
      if conditions for this class are met.

   ``(min-width)``, ``(max-width)``
      Minimum and maximum width of *parent element* required to apply given class.

   ``'self'``
      If this is specified as third parameter, then ``(min-width)`` and ``(max-width)`` are compared against the block
      in context, not its parent element.


scripts/settings.js
/*
 * Put global configuration options here.
 */
define('settings', {
    // Enable or disable jQuery 'no conflict' mode. Defaults to `true`.
    'JQUERY_NO_CONFLICT': true,

    'VIEWPORTS': {
        '320': [0, 479],
        '480': [480, 719],
        '720': [720, 959],
        '960': [960, 1279],
        '1280': [1280, 1599],
        '1600': [1600, 1919],
        '1920': [1920, Infinity]
    },
    // Uncomment this to disable 'viewport-portrait' and 'viewport-landscape' classes added automatically to body.
    // 'VIEWPORT_ORIENTATION_CLASS': false,

    'SMART_BLOCKS': {
        // Example smart block configuration.
        // Initial element selector.
        // '.block': {
        //     //Additional classes that will be applied when width of parent is between given numbers.
        //     'small': [0, 299],
        //     'medium': [300, 699],
        //     'big': [700, Infinity],
        // },
        // '.self-block': {
        //    // Additional classes that will be applied when width of element is between given numbers.
        //    'small': [0, 299, 'self'],
        //    'medium': [300, 699, 'self'],
        //    'big': [700, Infinity, 'self'],
        //}
    }
});
